lib/web_ui/lib/src/engine/ulps.dart

```diff
@@ -0,0 +1,194 @@
+}
+
+// Equality using the same error term as between
+bool almostBequalUlps(double a, double b) {
+  const int kUlpsEpsilon = 2;
+  return equalUlps(a, b, kUlpsEpsilon, kUlpsEpsilon);
+}
+
+bool almostPequalUlps(double a, double b) {
+  const int kUlpsEpsilon = 8;
+  return equalUlps(a, b, kUlpsEpsilon, kUlpsEpsilon);
+}
+
+bool almostDequalUlps(double a, double b) {
+  const int kUlpsEpsilon = 16;
+  return equalUlps(a, b, kUlpsEpsilon, kUlpsEpsilon);
+}
+
+bool approximatelyEqual(double ax, double ay, double bx, double by) {
+  if (approximatelyEqualT(ax, bx) && approximatelyEqualT(ay, by)) {
+    return true;
+  }
+  if (!roughlyEqualUlps(ax, bx) || !roughlyEqualUlps(ay, by)) {
+    return false;
+  }
+  final double dx = (ax - bx);
+  final double dy = (ay - by);
+  double dist = math.sqrt(dx * dx + dy * dy);
+  double tiniest = math.min(math.min(math.min(ax, bx), ay), by);
+  double largest = math.max(math.max(math.max(ax, bx), ay), by);
+  largest = math.max(largest, -tiniest);
+  return almostDequalUlps(largest, largest + dist);
+}
+
+// Use this for comparing Ts in the range of 0 to 1. For general numbers (larger and smaller) use
+// AlmostEqualUlps instead.
+bool approximatelyEqualT(double t1, double t2) {
+  return approximatelyZero(t1 - t2);
+}
+
+bool approximatelyZero(double value) => value.abs() < kFltEpsilon;
+
+bool roughlyEqualUlps(double a, double b) {
+  const int UlpsEpsilon = 256;
+  const int DUlpsEpsilon = 1024;
+  return equalUlps(a, b, UlpsEpsilon, DUlpsEpsilon);
+}
+
+bool dEqualUlpsEpsilon(double a, double b, int epsilon) {
+  int aBits = floatAs2sCompliment(a);
+  int bBits = floatAs2sCompliment(b);
+  // Find the difference in ULPs.
+  return aBits < bBits + epsilon && bBits < aBits + epsilon;
+}
+
+bool almostDequalUlpsDouble(double a, double b) {
+  if (a.abs() < kScalarMax && b.abs() < kScalarMax) {
+    return almostDequalUlps(a, b);
+  }
+  return (a - b).abs() / math.max(a.abs(), b.abs()) < kFltEpsilon * 16;
+}
+
+const double kFltEpsilon = 1.19209290E-07; // == 1 / (2 ^ 23)
+const double kDblEpsilon = 2.22045e-16;
+const double kFltEpsilonCubed = kFltEpsilon * kFltEpsilon * kFltEpsilon;
+const double kFltEpsilonHalf = kFltEpsilon / 2;
+const double kFltEpsilonDouble = kFltEpsilon * 2;
+const double kFltEpsilonOrderableErr = kFltEpsilon * 16;
+const double kFltEpsilonSquared = kFltEpsilon * kFltEpsilon;
+// Use a compile-time constant for FLT_EPSILON_SQRT to avoid initializers.
+// A 17 digit constant guarantees exact results.
+const double kFltEpsilonSqrt = 0.00034526697709225118; // sqrt(kFltEpsilon);
+const double kFltEpsilonInverse = 1 / kFltEpsilon;
+const double kDblEpsilonErr = kDblEpsilon * 4;
+const double kDblEpsilonSubdivideErr = kDblEpsilon * 16;
+const double kRoughEpsilon = kFltEpsilon * 64;
+const double kMoreRoughEpsilon = kFltEpsilon * 256;
+const double kWayRoughEpsilon = kFltEpsilon * 2048;
+const double kBumpEpsilon = kFltEpsilon * 4096;
```

Comment:
There's a bunch of multiples of these Epsilons that seem unused. Do we want to expose them?

Reply:
To be used shortly for path.combine implementation.